### PR TITLE
v1.0 web

### PR DESCRIPTION
**Pull request by ksyeo1010**
- add koala web binding
- add cypress unit + perf test
- add web workflow

markdown check fails because there is no main/demo/web path.

### CODE REVIEW

**Review by ErisMik — APPROVED** [review]

File: binding/web/src/koala.ts [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+
+    return Koala._init(
+      accessKey,
+      processCallback,
+      modelPath,
+      options
+    );
+  }
+
+  public static async _init(
```
[the remark this answers is not in the document]
Comment (ksyeo1010) [reply binding/web/src/koala.ts:218]: its for `KoalaWorker` as well.

File: .github/workflows/web-codestyle.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+on:
+  workflow_dispatch:
+  push:
+    branches: [ main ]
+    paths:
+      - '**/web/*.js'
+      - '**/web/*.ts'
+      - '.github/workflows/web-codestyle.yml'
+  pull_request:
+    branches: [ main ]
```
Comment (laves) [inline .github/workflows/web-codestyle.yml:12]:
```suggestion
    branches: [ main, 'v[0-9]+.[0-9]+' ]
```

File: .github/workflows/web-perf.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  workflow_dispatch:
+  push:
+    branches: [ main ]
+    paths:
+      - 'lib/wasm/**'
+      - 'lib/common/**'
+      - 'binding/web/test/koala_perf.test.ts'
+      - '.github/workflows/web-perf.yml'
+  pull_request:
+    branches: [ main ]
```
Comment (laves) [inline .github/workflows/web-perf.yml:13]: Needed for releases
```suggestion
    branches: [ main, 'v[0-9]+.[0-9]+' ]
```

File: .github/workflows/web-perf.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+name: Web Performance
+
+on:
+  workflow_dispatch:
+  push:
+    branches: [ main ]
+    paths:
+      - 'lib/wasm/**'
+      - 'lib/common/**'
+      - 'binding/web/test/koala_perf.test.ts'
```
Comment (laves) [inline .github/workflows/web-perf.yml:10]: `resources/audio_samples/**` too (incase our test audio changes)

File: .github/workflows/web-perf.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      - '.github/workflows/web-perf.yml'
+  pull_request:
+    branches: [ main ]
+    paths:
+      - 'lib/wasm/**'
+      - 'lib/common/**'
+      - 'binding/web/test/koala_perf.test.ts'
```
Comment (laves) [inline .github/workflows/web-perf.yml:17]: `resources/audio_samples/**`

File: .github/workflows/web-perf.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        python-version: [3.8]
+        node-version: [lts/*]
+        include:
+          - python-version: 3.8
+            procPerformanceThresholdSec: 1.2
```
Comment (laves) [inline .github/workflows/web-perf.yml:34]: do we need python?

File: .github/workflows/web.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  run:
+    working-directory: binding/web
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        python-version: [3.8]
```
Comment (laves) [inline .github/workflows/web.yml:36]: are we using this?

File: README.md [hunk 8]
```diff
@@ -22,9 +22,10 @@ Koala is an on-device noise suppression engine. Koala is:
     - [Demos](#demos)
         - [Python](#python-demos)
         - [C](#c-demos)
+        - [Web](#web-demo)
     - [SDKs](#sdks)
         - [Python](#python)
-        - [C](#c)
+        - [Web](#web)
```
Comment (laves) [inline README.md:28]: looks like you removed the link to the C section?

File: README.md [hunk 9]
```diff
@@ -157,6 +176,50 @@ Finally, when done be sure to release the acquired resources:
...
+import koalaParams from "${PATH_TO_BASE64_KOALA_PARAMS}";
+
+function processCallback(enhancedPcm) {
+  // do something with enhancedPcm
+}
+
+function getAudioData(): Int16Array {
+... // function to get audio data
+  return new Int16Array();
+}
```
Comment (laves) [inline README.md:206]: be more clear to call it "AudioFrame"
```suggestion
function getAudioFrame(): Int16Array {
  // ... function to get a frame of audio
}
```

File: README.md [hunk 10]
```diff
@@ -157,6 +176,50 @@ Finally, when done be sure to release the acquired resources:
...
+
+const koala = await KoalaWorker.create(
+  "${ACCESS_KEY}",
+  processCallback,
+  { base64: koalaParams },
+);
+
+await koala.reset();
+for (;;) {
+    await koala.process(getAudioData());
```
Comment (laves) [inline README.md:216]:
```suggestion
    await koala.process(getAudioFrame());
```

File: binding/web/README.md [hunk 11]
```diff
@@ -0,0 +1,181 @@
+# Koala Binding for Web
+
+## Koala Speech-to-Text Engine
```
Comment (laves) [inline binding/web/README.md:3]:
```suggestion
## Koala Noise Suppression Engine
```

File: binding/web/README.md [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+#### Init options
+
+Set `processErrorCallback` to handle errors if an error occurs while enhancing audio.
+
+```typescript
+// Optional
+const options = {
+  processErrorCallback: (error) => {}
+}
+```
```
Comment (laves) [inline binding/web/README.md:110]: I think you should just merge the contents of this section into the one below. There are a bit too many sections in this README

File: binding/web/src/koala_worker.ts [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+                    // eslint-disable-next-line no-console
+                    console.error(ev.data.message);
+                  }
+                  break;
+                default:
+                  // @ts-ignore
+                  processErrorCallback(`Unrecognized command: ${event.data.command}`);
+              }
+            };
+            resolve(new KoalaWorker(worker, event.data.version, event.data.frameLength, event.data.sampleRate, event.data.delaySample));
```
Comment (laves) [inline binding/web/src/koala_worker.ts:157]: can you spread the args to their own lines to make diffs easier?

File: binding/web/src/koala_worker.ts [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+      wasm: this._wasm,
+      wasmSimd: this._wasmSimd,
+    });
+
+    return returnPromise;
+  }
+
+  /**
+   * Processes a frame of audio in a worker.
+   * The transcript result will be supplied with the callback provided when initializing the worker either
```
Comment (laves) [inline binding/web/src/koala_worker.ts:184]: no transcript from Koala

**Review by laves — CHANGES_REQUESTED** [review]

**Review by laves — APPROVED** [review]

**Comment by laves** [comment]
go ahead and release pkgs

**Pull request merged**